### PR TITLE
BLD, TST: refactor test to use meson not setup.py, improve spin test -m ...

### PR DESCRIPTION
**Pull request by mattip**
This PR has two parts:
- Add `spin test -m full` which will run the full test suite
- Refactor `random/tests/test_extending.py` to use meson

Marking as draft to also refactor `numpy.testing.extbuild` to use meson as well.

### CODE REVIEW

**Comment by mattip** [comment]
Any ideas how to avoid `pip install`ing a package in a requirements file on pyodide? This works (will skip installing ninja) on ubuntu
```
ninja ; sys_platform != "linux"
```

but [this](https://github.com/numpy/numpy/pull/24153/files#diff-43f6847f9f058bf2ba296a9a3554506628b2dc79c663d4f7c266ba5589d7d618R6) seems to [not work](https://github.com/numpy/numpy/actions/runs/5529345244/jobs/10087280829#step:8:50) on pyodide

```
ninja ; sys_platform != "emscripten"
```

even though the `sys.platform` [is `emscripten`](https://github.com/numpy/numpy/actions/runs/5529345244/jobs/10087280829#step:8:32)

**Comment by rgommers** [comment]
I think `ninja` should be packaged in Pyodide, and until that happens I'd just skip the tests for `IS_WASM`. Cc @hoodmane for visibility/opinion.

**Comment by mattip** [comment]
The relevant tests are skipped. The problem is that `pip install -r test_requirements.txt` is failing on pyodide (before testing begins), since ninja cannot be installed. I could remove it from `test_requirements.txt` and install manually in all the other CI runs, but that seems wrong.

**Comment by rgommers** [comment]
Ah, that makes sense. It should already be present in most all CI runs (except for the ones with build isolation perhaps), because it's needed for building. It falls in the same category as compilers, pkg-config, etc. - I'd say it should not be in `test_requirements.txt` because `ninja` is not a Python package and you'd preferably get it from your system package manager (it being packaged on PyPI is a hack; meson-python has an elaborate workaround to only install it if it's not already present on the system).

**Comment by rgommers** [comment]
tl;dr this is an Ubuntu job, so just install it with `apt`?

**Comment by mattip** [comment]
> this is an Ubuntu job, so just install it with apt?

Kind of. "this" is also all wheel builds, which build in an isolated environment and then test.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -32,42 +32,6 @@ stages:
...
-            python3.9 runtests.py --show-build-log --cpu-baseline=native --cpu-dispatch=none \
-            --debug-info --mode=full -- -rsx --junitxml=junit/test-results.xml
-      displayName: 'Run native baseline Build / Tests'
-    - task: PublishTestResults@2
-      condition: succeededOrFailed()
-      inputs:
-        testResultsFiles: '**/test-*.xml'
-        failTaskOnFailedTests: true
-        testRunTitle: 'Publish test results for baseline/native'
-
```
Comment (mattip) [inline azure-pipelines.yml:70]: Delete this build job, we have a python3.9 linux job in the github build_test action.

**Comment by rgommers** [comment]
> So I still did `pip install` in places where I was too lazy to work out an alternative.

That sounds fine to me.

**Comment by mattip** [comment]
Looking more closely at the ninja PyPI package, it seems like a valid wheel solution to provide a binary for all platforms. It is a shame I needed to add it everywhere `pip install -r test_requirements.txt` is used instead of putting this inside `test_requirements.txt`
```
ninja ; sys_platform != "emscripten"
```

**Comment by mattip** [comment]
This is ready for review:
- move cython and c-extension compilation to meson
- add a c++ cython build test to the existing ones using the `nprandom` static library 
- remove a python3.9 azure job
- add a `spin -m full` command to run slow tests

**Comment by hoodmane** [comment]
On the topic of
```
ninja ; sys_platform != "emscripten"
```
I've run into the same problem while working on https://github.com/pypa/cibuildwheel/pull/1456 and I would like to fix it soon. There's a lot of complexity involved in convincing pip to behave correctly since it needs to run on the host linux Python but install stuff for emscripten Python. I think the correct way to fix this is to use `PYTHONPATH` to add customized versions of the `platform.py` and `sysconfig.py` but it is a bit tricky to get right.

File: tools/travis-test.sh [hunk 2]
```diff
@@ -80,6 +80,13 @@ setup_base()
 
 run_test()
 {
+  # meson has a hard dependency on ninja, and we need meson to build
```
Comment (rgommers) [inline tools/travis-test.sh:83]: minor: it'd be good not to repeat the same comment multiple times; can you either remove this or change to something like "see note in `pyproject.toml` for why ninja is installed as a test requirement"?

File: numpy/random/tests/test_extending.py [hunk 3]
```diff
@@ -39,55 +43,117 @@
...
     build_dir = tmp_path / 'random' / '_examples' / 'cython'
-    subprocess.check_call([sys.executable, 'setup.py', 'build', 'install',
-                           '--prefix', str(tmp_path / 'installdir'),
-                           '--single-version-externally-managed',
-                           '--record', str(tmp_path/ 'tmp_install_log.txt'),
-                          ],
-                          cwd=str(build_dir),
-                      )
+    # We don't want a wheel build, so do the steps in a controlled way
+    # The meson.build file is not copied as part of the build, so generate it
```
Comment (rgommers) [inline numpy/random/tests/test_extending.py:62]: This should not be necessary, you can install any file you want/need. Can you commit a regular `meson.build` file in `random/_examples/cython/` and edit `random/meson.build` to install it rather than the `_examples/cython/setup.py` file that's installed now?

**Review by rgommers — COMMENTED** [review]
This looks quite good, thanks Matti. I have only one significant comment.

**Comment by rgommers** [comment]
`time spin test` shows that the default test suite run goes from 7min03 to 2min47, which is nice for usability.

**Comment by mattip** [comment]
I addressed the review comments, CI is passing.

**Comment by rgommers** [comment]
> CI is passing.

It looks very red to me.

**Comment by rgommers** [comment]
I think you forgot to commit the new `meson.build` file?

**Comment by mattip** [comment]
> I think you forgot to commit the new meson.build file?

Thanks. Committed now, I guess I should refresh (f5) before jumping to conclusions.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks Matti!

**Comment by charris** [comment]
We may want to backport some of this to 1.26, so marking.

**Comment by charris** [comment]
> We may want to backport some of this to 1.26, so marking.

Seems we don't need it, so removing label.
